hbase-server/src/main/java/org/apache/hadoop/hbase/rsgroup/RSGroupInfoManagerImpl.java

```diff
@@ -356,9 +363,129 @@ public synchronized void removeServers(Set<Address> servers) throws IOException
     return RSGroupInfoList;
   }
 
-  @Override
-  public void refresh() throws IOException {
-    refresh(false);
+  private void waitUntilSomeProcsDone(Set<Long> pendingProcIds) {
+    int size = pendingProcIds.size();
+    while (!masterServices.isStopped()) {
+      for (Iterator<Long> iter = pendingProcIds.iterator(); iter.hasNext();) {
+        long procId = iter.next();
+        if (masterServices.getMasterProcedureExecutor().isFinished(procId)) {
+          iter.remove();
+        }
+      }
+      if (pendingProcIds.size() < size) {
+        return;
+      }
+      try {
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  private void waitUntilMasterStarted() {
+    while (!masterServices.isInitialized() && !masterServices.isStopped()) {
+      try {
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  private void migrate(Collection<RSGroupInfo> groupList, int maxConcurrency) {
+    waitUntilMasterStarted();
+    Set<Long> pendingProcIds = new HashSet<>();
+    for (RSGroupInfo groupInfo : groupList) {
+      if (groupInfo.getName().equals(RSGroupInfo.DEFAULT_GROUP)) {
+        continue;
+      }
+      SortedSet<TableName> failedTables = new TreeSet<>();
+      for (TableName tableName : groupInfo.getTables()) {
+        LOG.info("Migrating {} in group {}", tableName, groupInfo.getName());
+        TableDescriptor oldTd;
+        try {
+          oldTd = masterServices.getTableDescriptors().get(tableName);
+        } catch (IOException e) {
+          LOG.warn("Failed to migrate {} in group {}", tableName, groupInfo.getName(), e);
+          failedTables.add(tableName);
+          continue;
+        }
+        if (oldTd == null) {
+          continue;
+        }
+        if (oldTd.getRegionServerGroup().isPresent()) {
+          // either we have already migrated it or that user has set the rs group using the new
+          // code which will set the group directly on table descriptor, skip.
+          LOG.debug("Skip migrating {} since it is already in group {}", tableName,
+            oldTd.getRegionServerGroup().get());
+          continue;
+        }
+        TableDescriptor newTd = TableDescriptorBuilder.newBuilder(oldTd)
+          .setRegionServerGroup(groupInfo.getName()).build();
+        try {
+          pendingProcIds.add(
```

Comment:
For the user which used RSGroup feature a lot, when user restart master with new code, this migrate thing will reopen all regions... This is not friendly for user? Can we finish this job by a tool? Only use the tool to change the TableDescriptor first, then upgrade cluster to new version code.

Reply:
I think doing things automatically will be more friendly? As users may forget to user the tool and all the table configs will be ignored...

Reply:
Need a new issue to add document for this? And add a config for maxConcurrency? If user decide to upgrade, user can use this config to reduce the impact for availability.